pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_column_type.go

```diff
@@ -228,11 +240,130 @@ func handleGeneralColumnConversion(
+			// dropped by using dependency rules.
+			panic(scerrors.NotImplementedErrorf(t,
+				"backfilling during ALTER COLUMN TYPE for a column "+
+					"with a DEFAULT or ON UPDATE expression is not supported"))
 		}
 	})
 
-	// TODO(spilchen): Implement the general conversion logic in #127014
-	panic(scerrors.NotImplementedErrorf(t, "general alter type conversion not supported in the declarative schema changer"))
+	// We block any attempt to alter the type of a column that is a key column in
+	// the primary key. We can't use walkColumnDependencies here, as it doesn't
+	// differentiate between key columns and stored columns.
+	pk := mustRetrievePrimaryIndex(b, tbl.TableID)
+	for _, keyCol := range getIndexColumns(b.QueryByID(tbl.TableID), pk.IndexID, scpb.IndexColumn_KEY) {
+		if keyCol.ColumnID == col.ColumnID {
+			panic(sqlerrors.NewAlterColumnTypeColInIndexNotSupportedErr())
+		}
+	}
+
+	// TODO(#47137): Only support alter statements that only have a single command.
+	switch s := stmt.(type) {
+	case *tree.AlterTable:
+		if len(s.Cmds) > 1 {
+			panic(sqlerrors.NewAlterColTypeInCombinationNotSupportedError())
+		}
+	}
+
+	// In version 25.1, we introduced the necessary dependency rules to ensure the
+	// general path works. Without these rules, we encounter failures during the
+	// ALTER operation. To avoid this, we revert to legacy handling if not running
+	// on version 25.1.
+	// TODO(25.1): Update V24_3 here once V25_1 is defined.
+	if !b.EvalCtx().Settings.Version.ActiveVersion(b).IsActive(clusterversion.V24_3) {
+		panic(scerrors.NotImplementedErrorf(t,
+			"old active version; ALTER COLUMN TYPE requires backfill. Reverting to legacy handling"))
+	}
+
+	// TODO(#132936): Not yet supported in the DSC. Throwing an error to trigger
+	// fallback to legacy.
+	if newColType.Type.Family() == types.EnumFamily {
+		panic(scerrors.NotImplementedErrorf(t,
+			"backfilling during ALTER COLUMN TYPE for an enum column "+
+				"type is not supported"))
+	}
+
+	// Generate the ID of the new column we are adding.
+	newColID := b.NextTableColumnID(tbl)
+	newColType.ColumnID = newColID
+
+	// Set a computed expression for the new column that references the old column.
+	//
+	// The new column shares the old column's name, so we'll use a temporary name
+	// for the old column in the expression, stored in this variable.
+	colName := mustRetrieveColumnName(b, tbl.TableID, col.ColumnID)
+	expr, err := getComputeExpressionForBackfill(b, t, tn, tbl.TableID, colName.Name, newColType)
+	if err != nil {
+		panic(err)
+	}
+
+	// Drop the column that the new column will be replacing.
+	b.Drop(col)
+	b.Drop(colName)
+	b.Drop(oldColType)
+	handleDropColumnPrimaryIndexes(b, tbl, col)
+
+	// Add the new column. It will be identical to the column it is replacing,
+	// except the type will differ, and it will have a transient computed expression.
+	// This expression will reference the original column to facilitate the backfill.
+	spec := addColumnSpec{
```

Comment:
but the old column is still there until the first backfill what do you think if we say "set target status of old column to drop" and then in the new column text perhaps something like "after this backfill, the old column actually gets dropped"

i do think this is implied with your explanation further down so perhaps this is just a nit